ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixFactory.java

```diff
@@ -33,15 +40,73 @@
    */
   public HelixManager getZKHelixManager(String clusterName, String instanceName, InstanceType instanceType,
       String zkAddr) {
-    return HelixManagerFactory.getZKHelixManager(clusterName, instanceName, instanceType, zkAddr);
+    ManagerKey managerKey = new ManagerKey(clusterName, instanceName, instanceType, zkAddr);
+    return helixManagers.computeIfAbsent(managerKey,
+        k -> HelixManagerFactory.getZKHelixManager(clusterName, instanceName, instanceType, zkAddr));
   }
 
   /**
-   * Get a reference to a {@link HelixAdmin}
+   * Get a reference to a {@link HelixManager} and connect to it, if not already connected
+   * @param clusterName the name of the cluster for which the manager is to be gotten.
+   * @param instanceName the name of the instance on whose behalf the manager is to be gotten.
+   * @param instanceType the {@link InstanceType} of the requester.
    * @param zkAddr the address identifying the zk service to which this request is to be made.
-   * @return the constructed {@link HelixAdmin}.
+   * @return the constructed and connected {@link HelixManager}.
+   * @throws Exception if connecting failed.
+   */
+  public HelixManager getZkHelixManagerAndConnect(String clusterName, String instanceName, InstanceType instanceType,
+      String zkAddr) throws Exception {
+    HelixManager manager = getZKHelixManager(clusterName, instanceName, instanceType, zkAddr);
+    synchronized (manager) {
+      if (!manager.isConnected()) {
+        LOGGER.info("Connecting to HelixManager at {}", zkAddr);
+        manager.connect();
+        LOGGER.info("Established connection to HelixManager at {}", zkAddr);
+      } else {
+        LOGGER.info("HelixManager at {} already connected", zkAddr);
+      }
+    }
```

Comment:
There should not be such a case in production since we would never start `HelixClusterManager` and `HelixFactory` in parallel. I put this here since there is nothing that strictly guarantees that will always be the case. If you feel that this is unneeded for now I could take it out.

We could move this into a Map.compute lambda, but I felt that this approach allowed for finer grained locking as we don't need the construction and connection to be atomic, we just need to make sure that after construction and adding to the map, only one thread does the check connected, connect sequence at a time.

Reply:
I am okay with this approach after more clarification. Let's keep it as it is.